trl/trainer/grpo_trainer.py

```diff
@@ -316,9 +325,35 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
             prompt_inputs["input_ids"] = prompt_inputs["input_ids"][:, -self.max_prompt_length :]
             prompt_inputs["attention_mask"] = prompt_inputs["attention_mask"][:, -self.max_prompt_length :]
 
-        # Generate completions
-        with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
-            prompt_completion_ids = unwrapped_model.generate(**prompt_inputs, generation_config=self.generation_config)
+        # Generate completions using either vLLM or regular generation
+        if self.args.use_vllm:
+            # First, have main process load weights if needed
+            if self.state.global_step != self._last_loaded_step:
+                if self.accelerator.is_main_process:
+                    self.vllm_client.load_weights(self.accelerator.unwrap_model(model).state_dict())
+                self._last_loaded_step = self.state.global_step
+                self.accelerator.wait_for_everyone() # Wait for main process to finish loading weights
+
+            # Set up sampling parameters
+            sampling_params = {"n": self.args.num_generations, "temperature": self.args.temperature}
+
+            # Get completions from vLLM for all prompts
+            with self.accelerator.main_process_first():  # Why it doesn't work without this?
```

Comment:
Presumably the answer to your question is that without this you get a race condition with the other processes? What kind of error do you get?

Reply:
so there is no error I believe but it could be that the completions come from an "older" model, so we first get the main processor to update the server's model weights and then use it...

Reply:
For context, I run it with two processes.
I don't get an error, here, but for some reason, the one the two processes get an empty list for answer from the server.

Reply:
Can you share the full commands you're running (both the `trl vlm-serve` and the subsequent train command?